packages/gatsby/src/internal-plugins/dev-404-page/raw_dev-404-page.js

```diff
@@ -11,13 +12,16 @@ class Dev404Page extends React.Component {
 
   constructor(props) {
     super(props)
-    const { data } = this.props
+    const { data, location } = this.props
     const pagePaths = data.allSitePage.nodes.map(node => node.path)
+    const urlState = queryString.parse(location.hash)
+
+    const initialPagePathSearchTerms = urlState.q ? urlState.q : ``
     this.state = {
       showCustom404: false,
       initPagePaths: pagePaths,
       pagePaths: pagePaths,
```

Comment:
Oh, we also need to add filtering here, because otherwise we will still show every page

Reply:
True.